File: frontend/lib/library/courts.ts
// The 12 court buckets of the /judgments wing — «الجهة القضائية».
//
// ⚠ `shared/library/courts.py` IS THE SOURCE OF TRUTH. This file is a MIRROR of
// its `_COURTS` dict: the slug keys, the Arabic labels, and — critically — the
// INSERTION ORDER, which is corpus-volume order and IS the browse order.
// Alphabetical would bury المحكمة التجارية (20,335 rows) under المحكمة العامة
// (69). The two files must be edited in LOCKSTEP, in the same commit; the
// Python module additionally owns the raw `cases.court` → bucket variant map
// (the query predicate), which has no reason to exist in the browser and is
// deliberately NOT mirrored here.
//
// This is the same relationship `lib/library/court-levels.ts` has with
// `agents/deep_search_v4/shared/court_levels.py`: pure data + pure string
// helpers, no fetching, no client state, safe in both graphs.
//
// ⚠ THE SLUGS ARE ARABIC, and that is a deliberate departure from the
// «structural path segments are Latin» rule in `.claude/plans/library_sectors.md`
// (D4). That rule's justification was SEO neutrality and the whole /judgments
// wing is `noindex, nofollow` behind the PDPL gate, so there is no SEO to be
// neutral about — the user asked for Arabic explicitly. Two consequences:
//
//   1. A Next dynamic-route param arrives PERCENT-ENCODED for non-ASCII
//      (`%D8%A7%D9%84…`). Every entry point runs it through
//      `normalizeCourtSlug()` BEFORE comparing it to the vocabulary — a raw
//      `params.court` never matches `COURT_LABELS` on its own.
//   2. Anything sent onward is sent DECODED exactly once: the backend takes the
//      court as a QUERY param (`?court=…`), where `URLSearchParams` does the one
//      encode; hrefs carry the raw Arabic (the established pattern for Arabic
//      document slugs — `JudgmentCard` links `/judgments/${item.slug}`).
//      Encoding a param that arrived encoded produces `%25D8…`, which 404s.
//
// Counts, feeds and the raw-variant lists live server-side. Nothing here reads
// the corpus.

/**
 * `[slug, label]` for the 12 buckets, in the Python module's insertion order.
 *
 * One tuple per bucket rather than two parallel literals ON PURPOSE: the slug
 * and its label are the same Arabic text differing only by the hyphens, and
 * keeping them apart in an RTL editor is how a mismatched pair gets shipped.
 */
const COURTS = [
  // 20,335 rows
  ["المحكمة-التجارية", "المحكمة التجارية"],
  // 2,625
  ["ديوان-المظالم-تجارية", "ديوان المظالم — الدائرة التجارية"],
  // 2,281 — the untyped ZATCA residual (the LARGEST tax bucket; see the Python
  // module's docstring: the tax-type split only reaches 54% of that corpus).
  ["اللجان-الضريبية-عام", "اللجان الضريبية والزكوية — عام"],
  // 1,879
  ["ديوان-المظالم-إدارية", "ديوان المظالم — الدائرة الإدارية"],
  // 1,622
  ["لجان-ضريبة-القيمة-المضافة", "لجان ضريبة القيمة المضافة"],
  // 1,063
  ["لجان-ضريبة-الدخل-والزكاة", "لجان ضريبة الدخل والزكاة"],
  // 225
  ["لجان-التأمين", "لجان الفصل في المنازعات التأمينية"],
  // 165
  ["ديوان-المظالم-جزائية", "ديوان المظالم — الدائرة الجزائية"],
  // 125
  ["المحكمة-العليا", "المحكمة العليا"],
  // 106
  ["محكمة-الاستئناف", "محكمة الاستئناف"],
  // 69
  ["المحكمة-العامة", "المحكمة العامة"],
  // 35 — the entire labour corpus. Thin on purpose: the section is honest and
  // stands as a visible marker that labour judgments need sourcing.
  ["المحكمة-العمالية", "المحكمة العمالية"],
] as const;

/** The closed slug vocabulary, as a type. */
export type CourtSlug = (typeof COURTS)[number][0];

/** The 12 slugs in browse (corpus-volume) order. Never re-sort. */
export const COURT_ORDER: readonly CourtSlug[] = COURTS.map(([slug]) => slug);

/** slug → Arabic display label. This is the H1 and the switcher label. */
export const COURT_LABELS: Record<CourtSlug, string> = Object.fromEntries(
  COURTS,
) as Record<CourtSlug, string>;

/**
 * Segments under `/judgments/courts/` that are NOT court slugs.
 *
 * `/judgments/courts/page/{n}` must never resolve as a court in either
 * direction. Next resolves the static `page` segment first, and
 * `shared/library/courts.py` refuses the same two names server-side — two
 * layers, one vocabulary (mirrors `RESERVED_COURT_SLUGS` there).
 */
export const RESERVED_COURT_SLUGS: ReadonlySet<string> = new Set([
  "page",
  "mine",
]);

/**
 * Decode a `[court]` route param exactly ONCE and trim it.
 *
 * Next hands a non-ASCII dynamic param percent-encoded, so this is what turns
 * `%D8%A7%D9%84%D9%85%D8%AD%D9%83%D9%85%D8%A9-…` back into a slug the closed
 * vocabulary can be asked about. Malformed input (a lone `%`) throws inside
 * `decodeURIComponent`; that is not a crash, it is simply not one of the 12, so
 * the raw value falls through and fails validation a line later.
 *
 * Same decode-first normalisation as `encodeSlug()` in `lib/library/api.ts`,
 * minus the re-encode: the court reaches the backend as a QUERY param, where
 * `URLSearchParams` performs the single encode. Encoding here as well is the
 * `%25…` double-encode trap.
 */
export function normalizeCourtSlug(raw: string): string {
  let decoded = raw;
  try {
    decoded = decodeURIComponent(raw);
  } catch {
    decoded = raw;
  }
  return decoded.trim();
}

/** True when `value` is one of the 12 — pass it through `normalizeCourtSlug` first. */
export function isCourtSlug(value: string): value is CourtSlug {
  if (RESERVED_COURT_SLUGS.has(value)) return false;
  return value in COURT_LABELS;
}

/**
 * Arabic label for a court slug, or `null` when it is not one of the 12.
 *
 * Mirrors `courts.court_for_slug()`, including its contract: `null` means 404
 * and the caller MUST NOT spend a backend round-trip on it.
 */
export function courtLabel(slug: string): string | null {
  const normalized = normalizeCourtSlug(slug);
  return isCourtSlug(normalized) ? COURT_LABELS[normalized] : null;
}

/** `/judgments/courts/{slug}` — page 1 of one court section. */
export function courtPath(slug: string): string {
  return `/judgments/courts/${slug}`;
}

/** `/judgments/courts/{slug}/page/{n}` — a deep page of one court section. */
export function courtPagePath(slug: string, page: number): string {
  return `${courtPath(slug)}/page/${page}`;
}

/** H1 / `<title>` for a court section. The label is always the Arabic one. */
export function courtHeading(label: string): string {
  return `أحكام ${label}`;
}

/**
 * The facet's own name. **«الجهة القضائية», never «نوع المحكمة».**
 *
 * `المحكمة العليا` and `محكمة الاستئناف` are court LEVELS that leak into the
 * `court` column on the وزارة العدل feed, so those two buckets sit alongside the
 * existing درجة المحكمة chips. The two axes compose; calling this one «نوع
 * المحكمة» would make the pair read as a contradiction.
 */
export const COURT_FACET_LABEL = "الجهة القضائية";

/** One row of `GET /public/library/judgments/courts`, as the grid needs it. */
export interface CourtCountRow {
  slug: string;
  label?: string | null;
  count?: number | null;
}

/** One tile in the court switcher. `count` is null when the endpoint is down. */
export interface CourtNavItem {
  slug: CourtSlug;
  label: string;
  count: number | null;
}

/**
 * The 12 switcher tiles, counts attached where the API supplied them.
 *
 * ⚠ ORDER AND MEMBERSHIP COME FROM THE MIRROR, COUNTS COME FROM THE SERVER —
 * and that split is the point. The endpoint already returns browse order, so the
 * two agree by construction; iterating the mirror instead means (a) the switcher
 * still renders all 12 links when the counts call soft-fails, and (b) every href
 * is guaranteed to be a route that exists. A slug the server knows and this file
 * does not has no page to link to, so it is skipped — which is precisely the
 * drift the lockstep rule at the top exists to prevent, and it degrades to a
 * missing tile rather than a 404 in the reader's face.
 */
export function courtNavItems(
  rows?: readonly CourtCountRow[] | null,
): CourtNavItem[] {
  const counts = new Map<string, number>();
  for (const row of rows ?? []) {
    if (typeof row.count === "number") counts.set(row.slug, row.count);
  }
  return COURT_ORDER.map((slug) => ({
    slug,
    label: COURT_LABELS[slug],
    count: counts.get(slug) ?? null,
  }));
}

File: shared/library/courts.py
"""The court slug map — ONE place where a URL segment meets a raw ``cases.court`` string.

`public.cases.court` is free text straight from the source feeds: **30 distinct
values on 30,531 rows**, where the same judicial body appears under many spellings
that differ only by CITY («… في مدينة جدة» / «… في مدينة الرياض» / «… في مدينة
الدمام»). This module is the digest: it collapses those 30 raw strings into **12
browsable buckets** and is the ONLY normalizer for court names in the codebase.

Two rules the user set (2026-08-08):
  * **City never distinguishes a bucket.** «لجان الفصل في الرياض» and «لجان الفصل
    في جدة» are one court.
  * **Tax type DOES distinguish** — ZATCA committees split into ضريبة القيمة
    المضافة vs ضريبة الدخل والزكاة — and so do ديوان المظالم's دوائر
    (تجارية / إدارية / جزائية).

⚠ **THE TAX SPLIT ONLY REACHES 54% OF THE TAX CORPUS.** 2,281 of the 4,966 ZATCA
rows carry court strings that name no tax type at all («اللجنة الابتدائية الأولى»,
«اللجنة الاستئنافية», «لجنة الفصل الضريبي/الزكوي»). There is nothing in the field
to split them on, so they land in ``اللجان-الضريبية-عام``. That residual is the
LARGEST tax bucket and holds the best content in the feed (1,687 rows with
citation mesh — more than VAT and income combined). Splitting it further means
classifying from the judgment body, which is an LLM pass, not a string rule.

ORDER IS MEANINGFUL. ``COURTS`` is ordered by corpus volume and that insertion
order IS the browse order — alphabetical would bury المحكمة التجارية (20,335 rows)
under المحكمة العامة (69).

SLUGS ARE ARABIC here, unlike :mod:`shared.library.sectors`. `library_sectors.md`
D4 says structural path segments are Latin, and that rule's justification was SEO
neutrality — but the whole `/judgments` wing is ``noindex`` behind the PDPL gate,
so there is no SEO to be neutral about. The user asked for Arabic explicitly
(«judgment/ديوان المظالم»). Reversing the decision means editing only the slug
keys below; nothing else in the stack reads the slug's language.

DRIFT BEHAVIOUR — deliberate, and the same contract as ``sectors.py``: a raw court
value that no bucket claims is LOGGED and simply has no section page. It never
raises at import time (a pipeline-side re-ingest must not be able to crash the
backend's boot) and the judgment stays reachable via the unfiltered `/judgments`
hub and via search. ``backend/tests/test_library_courts.py`` asserts the live
distinct set against this map, so drift is caught in CI.

Row counts in the comments were measured 2026-08-08 and are documentation only —
nothing reads them.
"""
from __future__ import annotations

import logging
from typing import Iterable, Optional

logger = logging.getLogger(__name__)

# ─── Source feeds ─────────────────────────────────────────────────────────────
# The coarse axis. `legal_domains` and `date_gregorian` are populated PER FEED,
# not per row: every وزارة العدل row has both, every ديوان المظالم / ZATCA /
# insurance row has NEITHER. Anything that reasons about sector coverage or date
# ordering must branch on the feed, not on the individual court.
FEED_MOJ = "moj"                # وزارة العدل — 20,671 rows, sectors + dates
FEED_BOG = "board_of_grievances"  # ديوان المظالم — 4,669 rows, no sectors, NO citation mesh
FEED_ZATCA = "zatca"            # هيئة الزكاة والضريبة — 4,966 rows, no sectors, partial mesh
FEED_INSURANCE = "insurance"    # لجان التأمين — 225 rows, no sectors, NO citation mesh

# ─── The map ──────────────────────────────────────────────────────────────────
# slug → (Arabic label, raw `cases.court` variants, source feed)
# ORDERED BY CORPUS VOLUME. This order is the browse order.
_COURTS: dict[str, tuple[str, tuple[str, ...], str]] = {
    # 20,335
    "المحكمة-التجارية": (
        "المحكمة التجارية",
        ("التجارية", "التجارية الثالثة"),
        FEED_MOJ,
    ),
    # 2,625
    "ديوان-المظالم-تجارية": (
        "ديوان المظالم — الدائرة التجارية",
        ("ديوان المظالم — الدائرة التجارية",),
        FEED_BOG,
    ),
    # 2,281 — the untyped ZATCA residual. See the module docstring.
    "اللجان-الضريبية-عام": (
        "اللجان الضريبية والزكوية — عام",
        (
            "هيئة الزكاة والضريبة — اللجنة الاستئنافية الضريبية/الزكوية",
            "هيئة الزكاة والضريبة — اللجنة الابتدائية الأولى",
            "هيئة الزكاة والضريبة — اللجنة الاستئنافية",
            "هيئة الزكاة والضريبة — اللجنة الابتدائية الثانية",
            "هيئة الزكاة والضريبة — لجنة الفصل الضريبي/الزكوي",
            "هيئة الزكاة والضريبة — اللجنة الابتدائية الثالثة",
        ),
        FEED_ZATCA,
    ),
    # 1,879
    "ديوان-المظالم-إدارية": (
        "ديوان المظالم — الدائرة الإدارية",
        ("ديوان المظالم — الدائرة الإدارية",),
        FEED_BOG,
    ),
    # 1,622 — six variants differing only by city (جدة / الرياض / الدمام) + دائرة number.
    "لجان-ضريبة-القيمة-المضافة": (
        "لجان ضريبة القيمة المضافة",
        (
            "هيئة الزكاة والضريبة — الدائرة الأولى للفصل في مخالفات ومنازعات ضريبة القيمة المضافة في مدينة جدة",
            "هيئة الزكاة والضريبة — الدائرة الأولى للفصل في مخالفات ومنازعات ضريبة القيمة المضافة في مدينة الرياض",
            "هيئة الزكاة والضريبة — الدائرة الأولى للفصل في مخالفات ومنازعات ضريبة القيمة المضافة في مدينة الدمام",
            "هيئة الزكاة والضريبة — الدائرة الاستئنافية الأولى لمخالفات ومنازعات ضريبة القيمة المضافة والسلع الانتقائية في مدينة الرياض",
            "هيئة الزكاة والضريبة — الدائرة الثانية للفصل في مخالفات ومنازعات ضريبة القيمة المضافة في مدينة الرياض",
            "هيئة الزكاة والضريبة — الدائرة الثالثة للفصل في مخالفات ومنازعات ضريبة القيمة المضافة في مدينة الرياض",
        ),
        FEED_ZATCA,
    ),
    # 1,063 — income tax + the zakat/income appeal circuit (same subject matter).
    "لجان-ضريبة-الدخل-والزكاة": (
        "لجان ضريبة الدخل والزكاة",
        (
            "هيئة الزكاة والضريبة — الدائرة الأولى للفصل في مخالفات ومنازعات ضريبة الدخل في مدينة الرياض",
            "هيئة الزكاة والضريبة — الدائرة الأولى للفصل في مخالفات ومنازعات ضريبة الدخل في مدينة الدمام",
            "هيئة الزكاة والضريبة — الدائرة الثانية للفصل في مخالفات ومنازعات ضريبة الدخل في مدينة الرياض",
            "هيئة الزكاة والضريبة — الدائرة الأولى للفصل في مخالفات ومنازعات ضريبة الدخل في مدينة جدة",
            "هيئة الزكاة والضريبة — الدائرة الاستئنافية الأولى لمخالفات ومنازعات الزكاة وضريبة الدخل في مدينة الرياض",
            "هيئة الزكاة والضريبة — الدائرة الثالثة للفصل في مخالفات ومنازعات ضريبة الدخل في مدينة الرياض",
        ),
        FEED_ZATCA,
    ),
    # 225 — the whole insurance corpus. Separate from the tax لجان (user decision).
    "لجان-التأمين": (
        "لجان الفصل في المنازعات التأمينية",
        ("لجان الفصل في المنازعات والمخالفات التأمينية",),
        FEED_INSURANCE,
    ),
    # 165
    "ديوان-المظالم-جزائية": (
        "ديوان المظالم — الدائرة الجزائية",
        ("ديوان المظالم — الدائرة الجزائية",),
        FEED_BOG,
    ),
    # 125 — NOTE the second variant has TWO spaces after the dash. Copied verbatim.
    "المحكمة-العليا": (
        "المحكمة العليا",
        ("العليا", "العليا -  الهيئة الدائمة"),
        FEED_MOJ,
    ),
    # 106
    "محكمة-الاستئناف": ("محكمة الاستئناف", ("الاستئناف",), FEED_MOJ),
    # 69
    "المحكمة-العامة": ("المحكمة العامة", ("العامة",), FEED_MOJ),
    # 35 — the entire labour corpus. Shipped thin on purpose (user decision
    # 2026-08-08): the route is honest and stands as a visible marker that labour
    # judgments need SOURCING, which is a scraping job and not part of this work.
    "المحكمة-العمالية": ("المحكمة العمالية", ("العمالية",), FEED_MOJ),
}

# ─── Derived lookups ──────────────────────────────────────────────────────────
COURT_ORDER: list[str] = list(_COURTS)
"""The 12 court slugs in browse (volume) order."""

COURT_LABELS: dict[str, str] = {slug: label for slug, (label, _v, _f) in _COURTS.items()}
"""slug → Arabic display label. This is the H1 and the switcher label."""

COURT_VARIANTS: dict[str, tuple[str, ...]] = {
    slug: variants for slug, (_l, variants, _f) in _COURTS.items()
}
"""slug → the raw ``cases.court`` strings it claims. Feeds the ``in.()`` predicate."""

COURT_FEED: dict[str, str] = {slug: feed for slug, (_l, _v, feed) in _COURTS.items()}
"""slug → source feed. Branch on this for sector/date availability, never on the slug."""

COURT_SLUG_VOCAB: frozenset[str] = frozenset(_COURTS)
"""Closed slug vocabulary — validate against this BEFORE any DB work."""

# raw court string → slug. Built from the map so the two can never disagree.
_SLUG_BY_RAW: dict[str, str] = {
    raw: slug for slug, variants in COURT_VARIANTS.items() for raw in variants
}

# ─── Reserved segments ────────────────────────────────────────────────────────
# `/judgments/courts/page/{n}` must never resolve as a court, in either
# direction. Next resolves static segments first, but the BACKEND must refuse
# them too so the two namespaces can never collide. Same contract as
# ``sectors.RESERVED_SECTOR_SLUGS``.
RESERVED_COURT_SLUGS: frozenset[str] = frozenset({"page", "mine"})

# ─── Integrity check (log-and-continue, never raise) ──────────────────────────
_dupes = [
    raw
    for slug, variants in COURT_VARIANTS.items()
    for raw in variants
    if _SLUG_BY_RAW.get(raw) != slug
]
if _dupes:
    logger.error(
        "shared.library.courts: %d raw court value(s) are claimed by more than one "
        "bucket and will resolve to whichever was declared LAST: %s",
        len(_dupes),
        ", ".join(sorted(set(_dupes))),
    )
_reserved_collision = COURT_SLUG_VOCAB & RESERVED_COURT_SLUGS
if _reserved_collision:
    logger.error(
        "shared.library.courts: slug(s) collide with a reserved segment and will "
        "be unreachable: %s",
        ", ".join(sorted(_reserved_collision)),
    )


def court_for_slug(slug: Optional[str]) -> Optional[str]:
    """Arabic label for a court slug, or ``None`` if it is not one of the 12.

    Returns ``None`` for reserved segments even if a future edit added them to the
    map. Callers treat ``None`` as a 404 and MUST NOT issue a DB round-trip for it.
    """
    if not slug:
        return None
    slug = slug.strip()
    if slug in RESERVED_COURT_SLUGS:
        return None
    return COURT_LABELS.get(slug)


def variants_for_slug(slug: Optional[str]) -> Optional[tuple[str, ...]]:
    """Raw ``cases.court`` values a slug claims, or ``None`` if it is not a court.

    This tuple is what goes into the PostgREST ``in.()`` predicate. It is never a
    LIKE/regex: the vocabulary is closed and exact matching is what keeps the
    section's counts stable enough to stay OUT of the enumeration-oracle clamp.
    """
    if not slug:
        return None
    slug = slug.strip()
    if slug in RESERVED_COURT_SLUGS:
        return None
    return COURT_VARIANTS.get(slug)


def slug_for_court(raw_court: Optional[str]) -> Optional[str]:
    """Bucket slug for a raw ``cases.court`` string, or ``None`` if unclaimed.

    THE normalizer. Used to turn the court pill on a judgment card into a link and
    to bucket rows in the publish selector. An unclaimed value (today: the single
    empty-string row) renders as plain text rather than a broken link.
    """
    if not raw_court:
        return None
    return _SLUG_BY_RAW.get(raw_court.strip())


def feed_for_court(raw_court: Optional[str]) -> Optional[str]:
    """Source feed for a raw ``cases.court`` string, or ``None`` if unclaimed."""
    slug = slug_for_court(raw_court)
    return COURT_FEED.get(slug) if slug else None


def slugs_for_feed(feed: str) -> list[str]:
    """Court slugs belonging to one source feed, in browse order."""
    return [slug for slug in COURT_ORDER if COURT_FEED[slug] == feed]


def unclaimed_courts(raw_courts: Iterable[str]) -> list[str]:
    """Raw court values from the live corpus that no bucket claims.

    The CI drift check: pass ``select distinct court from cases`` and assert the
    result is the known-empty set. A new source feed shows up here first.
    """
    return sorted({c.strip() for c in raw_courts if c and c.strip() not in _SLUG_BY_RAW})


__all__ = [
    "FEED_MOJ",
    "FEED_BOG",
    "FEED_ZATCA",
    "FEED_INSURANCE",
    "COURT_ORDER",
    "COURT_LABELS",
    "COURT_VARIANTS",
    "COURT_FEED",
    "COURT_SLUG_VOCAB",
    "RESERVED_COURT_SLUGS",
    "court_for_slug",
    "variants_for_slug",
    "slug_for_court",
    "feed_for_court",
    "slugs_for_feed",
    "unclaimed_courts",
]
